guardian/newsletters-nx
# Dblatcher/s3 storage

## What does this change?

Implements a DraftStorage that reads and writes to an s3 test bucket rather instead of local memory. Local memory is still the default, but the storage instance used can be set by a an environment variable, USE_TEST_S3_STORAGE.

## How to test

Reversing the conditional in "apps/newsletters-api/src/services/storage/index.ts" before `npm run dev` will cause the API to use the s3 draft storage by default, IE:
```
const draftStore: DraftStorage = !isUsingTestS3Storage()
	? s3DraftStorage
	: inMemoryStorageInstance;
```
(TO DO - how to set environment variables on newsletters-api when running locally)

When running without getting credentials for the "developerPlayground" account from Janus, any API request for drafts (eg "http://localhost:3000/api/drafts") should fail with a 403 error.

When running with credentials, the UI and API should work as before, with any changes made being reflected in the test bucket and persisted between sessions - IE create a new draft using the wizard, and it should still be on the list after restarting the application.

## How can we measure success?

Will serve as a foundation/proof of concept for setting up buckets for the CODE and PROD application.

## Have we considered potential risks?

Before proceeding to use this in CODE or PROD, we should ask for help from DevX to make sure s3 is being accessed properly/securely. 
It make be that the s3 parameter (account, region, bucketName) should be environment variables or loaded from the s3 parameter store.
